File: plugins/lgtm/hooks/approve.py
#!/usr/bin/env python3
"""lgtm — PreToolUse hook: stamp "Looks Good To Me" unless denied.

Returns permissionDecision="allow" for tool calls that do not match any
permissions.deny rule found in user/project settings. Returns "deny" for
calls that confidently match a deny rule, and "ask" (defer to Claude Code's
own permission engine) when a deny rule's syntax isn't understood by this
matcher.

Settings sources scanned (all merged for deny rules):
  - ~/.claude/settings.json
  - $CLAUDE_PROJECT_DIR/.claude/settings.json
  - $CLAUDE_PROJECT_DIR/.claude/settings.local.json

Every decision is appended as a JSON line to:
  ${LGTM_LOG_FILE:-~/.claude/logs/lgtm.jsonl}
Set LGTM_LOG_FILE to an empty string to disable logging.
"""

from __future__ import annotations

import fnmatch
import json
import os
import re
import sys
from datetime import datetime, timezone
from pathlib import Path
from typing import Any
from urllib.parse import urlparse

FILE_PATH_KEYS = ("file_path", "path", "notebook_path")
FILE_TOOLS = {"Read", "Write", "Edit", "Glob", "NotebookEdit", "NotebookRead", "MultiEdit"}


def load_deny_rules() -> list[str]:
    paths: list[Path] = []
    paths.append(Path.home() / ".claude" / "settings.json")

    project_dir = os.environ.get("CLAUDE_PROJECT_DIR")
    if project_dir:
        project_root = Path(project_dir)
        paths.append(project_root / ".claude" / "settings.json")
        paths.append(project_root / ".claude" / "settings.local.json")

    rules: list[str] = []
    for p in paths:
        try:
            data = json.loads(p.read_text())
        except (FileNotFoundError, json.JSONDecodeError, OSError):
            continue
        perms = data.get("permissions") or {}
        deny = perms.get("deny") or data.get("deny") or []
        if isinstance(deny, list):
            rules.extend(r for r in deny if isinstance(r, str))
    return rules


RULE_RE = re.compile(r"^([A-Za-z0-9_]+)(?:\((.*)\))?$")


def parse_rule(rule: str) -> tuple[str, str | None] | None:
    m = RULE_RE.match(rule.strip())
    if not m:
        return None
    return m.group(1), m.group(2)


def tool_name_matches(rule_tool: str, tool_name: str) -> bool:
    if rule_tool == tool_name:
        return True
    # MCP server-prefix rule: "mcp__github" matches "mcp__github__create_issue", etc.
    if rule_tool.startswith("mcp__") and tool_name.startswith(rule_tool + "__"):
        return True
    return False


def path_matches(val: str, pattern: str) -> bool:
    """Match a path against a Claude Code-style permission glob.

    fnmatch alone doesn't honor `**` as recursive (it treats `**` as just `*`),
    so `Read(**/.env)` against bare `.env` would miss. We layer extra rules
    on top: `**/X` means "X at any depth, including root", `X/**` means
    "anything inside X/", and a bare basename matches anywhere in the path.
    """
    if fnmatch.fnmatch(val, pattern):
        return True

    if pattern.startswith("**/"):
        suffix = pattern[3:]
        if fnmatch.fnmatch(val, suffix):
            return True
        # Try every right-anchored subpath, so config/foo/.env still matches.
        parts = val.split("/")
        for i in range(len(parts)):
            if fnmatch.fnmatch("/".join(parts[i:]), suffix):
                return True

    if pattern.endswith("/**"):
        prefix = pattern[:-3]
        if val == prefix or val.startswith(prefix + "/"):
            return True

    # Bare basename pattern (no separator) matches if any path component matches.
    if "/" not in pattern and fnmatch.fnmatch(os.path.basename(val), pattern):
        return True

    return False


def match_rule(rule: str, tool_name: str, tool_input: dict[str, Any]) -> tuple[str, str]:
    """Return (status, reason). status ∈ {'match', 'no_match', 'unknown'}."""
    parsed = parse_rule(rule)
    if not parsed:
        return "unknown", f"unparseable rule: {rule!r}"
    rule_tool, rule_arg = parsed

    if not tool_name_matches(rule_tool, tool_name):
        return "no_match", ""

    if rule_arg is None:
        return "match", f"tool {tool_name} matches bare rule {rule!r}"

    if tool_name == "Bash":
        cmd = tool_input.get("command", "")
        if rule_arg.endswith(":*"):
            prefix = rule_arg[:-2]
            return ("match" if cmd.startswith(prefix) else "no_match",
                    f"command prefix {prefix!r} vs {cmd!r}")
        return ("match" if cmd == rule_arg else "no_match",
                f"command exact {rule_arg!r} vs {cmd!r}")

    if tool_name == "WebFetch":
        url = tool_input.get("url", "")
        if rule_arg.startswith("domain:"):
            target = rule_arg[len("domain:"):]
            host = urlparse(url).hostname or ""
            ok = host == target or host.endswith("." + target)
            return ("match" if ok else "no_match", f"host {host!r} vs domain {target!r}")
        return "unknown", f"unrecognized WebFetch arg syntax: {rule_arg!r}"

    if tool_name in FILE_TOOLS:
        for key in FILE_PATH_KEYS:
            val = tool_input.get(key)
            if isinstance(val, str):
                ok = path_matches(val, rule_arg)
                return ("match" if ok else "no_match", f"{key}={val!r} vs glob {rule_arg!r}")
        return "no_match", "no file path in tool_input"

    return "unknown", f"no matcher for tool {tool_name} with arg syntax"


def evaluate(tool_name: str, tool_input: dict[str, Any], deny_rules: list[str]) -> tuple[str, str]:
    ambiguous: list[str] = []
    for rule in deny_rules:
        status, why = match_rule(rule, tool_name, tool_input)
        if status == "match":
            return "deny", f"lgtm: matched deny rule {rule!r} — {why}"
        if status == "unknown":
            ambiguous.append(rule)
    if ambiguous:
        return "ask", (
            "lgtm: deferring to Claude Code's permission engine because "
            f"these deny rules could not be evaluated by the hook: {ambiguous}"
        )
    return "allow", "lgtm: no deny rule matched"


def resolve_log_path() -> Path | None:
    override = os.environ.get("LGTM_LOG_FILE")
    if override is not None:
        if override == "":
            return None
        return Path(override).expanduser()
    return Path.home() / ".claude" / "logs" / "lgtm.jsonl"


def _excerpt_input(tool_input: dict[str, Any]) -> dict[str, Any]:
    """Trim tool_input for logging so secrets/blobs don't bloat the log."""
    out = {}
    for k, v in tool_input.items():
        if isinstance(v, str) and len(v) > 200:
            out[k] = v[:200] + f"... ({len(v)} chars)"
        else:
            out[k] = v
    return out


def log_decision(tool_name: str, tool_input: dict[str, Any], decision: str, reason: str) -> None:
    log_path = resolve_log_path()
    if log_path is None:
        return
    try:
        log_path.parent.mkdir(parents=True, exist_ok=True)
        entry = {
            "ts": datetime.now(timezone.utc).isoformat(),
            "tool": tool_name,
            "decision": decision,
            "reason": reason,
            "input": _excerpt_input(tool_input),
            "cwd": os.environ.get("CLAUDE_PROJECT_DIR", ""),
        }
        with log_path.open("a") as f:
            f.write(json.dumps(entry) + "\n")
    except OSError:
        # Logging must never break tool execution.
        pass


def emit(decision: str, reason: str) -> None:
    out = {
        "hookSpecificOutput": {
            "hookEventName": "PreToolUse",
            "permissionDecision": decision,
            "permissionDecisionReason": reason,
        },
        "systemMessage": reason,
        "suppressOutput": True,
    }
    json.dump(out, sys.stdout)


def main() -> int:
    try:
        payload = json.load(sys.stdin)
    except json.JSONDecodeError:
        emit("ask", "lgtm: malformed payload, deferring")
        return 0

    tool_name = payload.get("tool_name", "")
    tool_input = payload.get("tool_input", {}) or {}

    deny_rules = load_deny_rules()
    decision, reason = evaluate(tool_name, tool_input, deny_rules)
    log_decision(tool_name, tool_input, decision, reason)
    emit(decision, reason)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: plugins/lgtm/hooks/approve.js
#!/usr/bin/env node
/**
 * lgtm — PreToolUse hook: stamp "Looks Good To Me" unless denied.
 *
 * Emits permissionDecision="allow" for tool calls that don't match any
 * permissions.deny rule from user/project settings. Emits "deny" when a
 * deny rule clearly matches, and "ask" (defer to Claude Code's own
 * permission engine) when a deny rule's syntax isn't understood here.
 *
 * Settings sources scanned (all merged):
 *   - ~/.claude/settings.json
 *   - $CLAUDE_PROJECT_DIR/.claude/settings.json
 *   - $CLAUDE_PROJECT_DIR/.claude/settings.local.json
 *
 * Decisions are appended as JSON lines to:
 *   ${LGTM_LOG_FILE:-~/.claude/logs/lgtm.jsonl}
 * Set LGTM_LOG_FILE="" to disable logging.
 *
 * Zero external dependencies — runs on whatever node ships with Claude Code.
 */

'use strict';

const fs = require('fs');
const os = require('os');
const path = require('path');

const FILE_TOOLS = new Set([
  'Read', 'Write', 'Edit', 'Glob', 'NotebookEdit', 'NotebookRead', 'MultiEdit',
]);
const FILE_PATH_KEYS = ['file_path', 'path', 'notebook_path'];
const RULE_RE = /^([A-Za-z0-9_]+)(?:\((.*)\))?$/;

function loadDenyRules() {
  const paths = [path.join(os.homedir(), '.claude', 'settings.json')];
  if (process.env.CLAUDE_PROJECT_DIR) {
    const root = process.env.CLAUDE_PROJECT_DIR;
    paths.push(path.join(root, '.claude', 'settings.json'));
    paths.push(path.join(root, '.claude', 'settings.local.json'));
  }

  const rules = [];
  for (const p of paths) {
    let data;
    try {
      data = JSON.parse(fs.readFileSync(p, 'utf-8'));
    } catch {
      continue;
    }
    const perms = data.permissions || {};
    const deny = perms.deny || data.deny || [];
    if (Array.isArray(deny)) {
      for (const r of deny) if (typeof r === 'string') rules.push(r);
    }
  }
  return rules;
}

/**
 * Translate an fnmatch-style glob to a regex and test it. Supports `*`, `?`,
 * `[abc]` character classes; escapes other regex metacharacters.
 */
function fnmatch(name, pattern) {
  let regex = '^';
  let i = 0;
  while (i < pattern.length) {
    const c = pattern[i];
    if (c === '*') {
      regex += '.*';
      i++;
    } else if (c === '?') {
      regex += '.';
      i++;
    } else if (c === '[') {
      const close = pattern.indexOf(']', i + 1);
      if (close === -1) {
        regex += '\\[';
        i++;
      } else {
        regex += pattern.slice(i, close + 1);
        i = close + 1;
      }
    } else if ('.+^${}()|\\'.indexOf(c) !== -1) {
      regex += '\\' + c;
      i++;
    } else {
      regex += c;
      i++;
    }
  }
  regex += '$';
  return new RegExp(regex).test(name);
}

/**
 * Path-glob match honoring `**` semantics that fnmatch alone doesn't give us:
 *   - `**\/X` matches X at any depth, including root.
 *   - `X/**` matches anything under X/.
 *   - bare basename patterns match any path component.
 */
function pathMatches(val, pattern) {
  if (fnmatch(val, pattern)) return true;

  if (pattern.startsWith('**/')) {
    const suffix = pattern.slice(3);
    if (fnmatch(val, suffix)) return true;
    const parts = val.split('/');
    for (let i = 0; i < parts.length; i++) {
      if (fnmatch(parts.slice(i).join('/'), suffix)) return true;
    }
  }

  if (pattern.endsWith('/**')) {
    const prefix = pattern.slice(0, -3);
    if (val === prefix || val.startsWith(prefix + '/')) return true;
  }

  if (!pattern.includes('/') && fnmatch(path.basename(val), pattern)) return true;

  return false;
}

function parseRule(rule) {
  const m = rule.trim().match(RULE_RE);
  if (!m) return null;
  return { tool: m[1], arg: m[2] !== undefined ? m[2] : null };
}

function toolNameMatches(ruleTool, toolName) {
  if (ruleTool === toolName) return true;
  // MCP server-prefix: "mcp__github" matches "mcp__github__create_issue", etc.
  if (ruleTool.startsWith('mcp__') && toolName.startsWith(ruleTool + '__')) return true;
  return false;
}

function matchRule(rule, toolName, toolInput) {
  const parsed = parseRule(rule);
  if (!parsed) return { status: 'unknown', why: `unparseable rule: ${JSON.stringify(rule)}` };
  const { tool: ruleTool, arg: ruleArg } = parsed;

  if (!toolNameMatches(ruleTool, toolName)) return { status: 'no_match', why: '' };

  if (ruleArg === null) {
    return { status: 'match', why: `tool ${toolName} matches bare rule ${JSON.stringify(rule)}` };
  }

  if (toolName === 'Bash') {
    const cmd = toolInput.command || '';
    if (ruleArg.endsWith(':*')) {
      const prefix = ruleArg.slice(0, -2);
      return {
        status: cmd.startsWith(prefix) ? 'match' : 'no_match',
        why: `command prefix ${JSON.stringify(prefix)} vs ${JSON.stringify(cmd)}`,
      };
    }
    return {
      status: cmd === ruleArg ? 'match' : 'no_match',
      why: `command exact ${JSON.stringify(ruleArg)} vs ${JSON.stringify(cmd)}`,
    };
  }

  if (toolName === 'WebFetch') {
    const url = toolInput.url || '';
    if (ruleArg.startsWith('domain:')) {
      const target = ruleArg.slice('domain:'.length);
      let host = '';
      try { host = new URL(url).hostname || ''; } catch { /* invalid URL */ }
      const ok = host === target || host.endsWith('.' + target);
      return {
        status: ok ? 'match' : 'no_match',
        why: `host ${JSON.stringify(host)} vs domain ${JSON.stringify(target)}`,
      };
    }
    return { status: 'unknown', why: `unrecognized WebFetch arg syntax: ${JSON.stringify(ruleArg)}` };
  }

  if (FILE_TOOLS.has(toolName)) {
    for (const key of FILE_PATH_KEYS) {
      const val = toolInput[key];
      if (typeof val === 'string') {
        const ok = pathMatches(val, ruleArg);
        return {
          status: ok ? 'match' : 'no_match',
          why: `${key}=${JSON.stringify(val)} vs glob ${JSON.stringify(ruleArg)}`,
        };
      }
    }
    return { status: 'no_match', why: 'no file path in tool_input' };
  }

  return { status: 'unknown', why: `no matcher for tool ${toolName} with arg syntax` };
}

function evaluate(toolName, toolInput, denyRules) {
  const ambiguous = [];
  for (const rule of denyRules) {
    const { status, why } = matchRule(rule, toolName, toolInput);
    if (status === 'match') {
      return { decision: 'deny', reason: `lgtm: matched deny rule ${JSON.stringify(rule)} — ${why}` };
    }
    if (status === 'unknown') ambiguous.push(rule);
  }
  if (ambiguous.length > 0) {
    return {
      decision: 'ask',
      reason: `lgtm: deferring to Claude Code's permission engine because these deny rules could not be evaluated by the hook: ${JSON.stringify(ambiguous)}`,
    };
  }
  return { decision: 'allow', reason: 'lgtm: no deny rule matched' };
}

function resolveLogPath() {
  const override = process.env.LGTM_LOG_FILE;
  if (override !== undefined) {
    if (override === '') return null;
    return override.startsWith('~/') ? path.join(os.homedir(), override.slice(2)) : override;
  }
  return path.join(os.homedir(), '.claude', 'logs', 'lgtm.jsonl');
}

function excerptInput(toolInput) {
  const out = {};
  for (const [k, v] of Object.entries(toolInput)) {
    if (typeof v === 'string' && v.length > 200) {
      out[k] = v.slice(0, 200) + `... (${v.length} chars)`;
    } else {
      out[k] = v;
    }
  }
  return out;
}

function logDecision(toolName, toolInput, decision, reason) {
  const logPath = resolveLogPath();
  if (!logPath) return;
  try {
    fs.mkdirSync(path.dirname(logPath), { recursive: true });
    const entry = {
      ts: new Date().toISOString(),
      tool: toolName,
      decision,
      reason,
      input: excerptInput(toolInput),
      cwd: process.env.CLAUDE_PROJECT_DIR || '',
    };
    fs.appendFileSync(logPath, JSON.stringify(entry) + '\n');
  } catch {
    // Logging must never break tool execution.
  }
}

function emit(decision, reason) {
  process.stdout.write(JSON.stringify({
    hookSpecificOutput: {
      hookEventName: 'PreToolUse',
      permissionDecision: decision,
      permissionDecisionReason: reason,
    },
    systemMessage: reason,
    suppressOutput: true,
  }));
}

async function readStdin() {
  return new Promise((resolve) => {
    let data = '';
    process.stdin.setEncoding('utf-8');
    process.stdin.on('data', (chunk) => { data += chunk; });
    process.stdin.on('end', () => resolve(data));
    process.stdin.on('error', () => resolve(data));
  });
}

(async function main() {
  let payload;
  try {
    payload = JSON.parse(await readStdin());
  } catch {
    emit('ask', 'lgtm: malformed payload, deferring');
    return;
  }

  const toolName = payload.tool_name || '';
  const toolInput = payload.tool_input || {};

  try {
    const denyRules = loadDenyRules();
    const { decision, reason } = evaluate(toolName, toolInput, denyRules);
    logDecision(toolName, toolInput, decision, reason);
    emit(decision, reason);
  } catch (e) {
    emit('ask', `lgtm: hook error: ${e.message}`);
  }
})();
